src/codegen/unwinding.cpp

```diff
@@ -410,23 +417,83 @@ static unw_word_t getFunctionEnd(unw_word_t ip) {
     return pip.end_ip;
 }
 
+static bool inASTInterpreterExecuteInner(unw_word_t ip) {
+    static unw_word_t interpreter_instr_end = getFunctionEnd((unw_word_t)interpreter_instr_addr);
+    return ((unw_word_t)interpreter_instr_addr <= ip && ip < interpreter_instr_end);
+}
+
+static bool inGeneratorEntry(unw_word_t ip) {
+    static unw_word_t generator_entry_end = getFunctionEnd((unw_word_t)generatorEntry);
+    return ((unw_word_t)generatorEntry <= ip && ip < generator_entry_end);
+}
+
+
+static inline unw_word_t get_cursor_reg(unw_cursor_t* cursor, int reg) {
+    unw_word_t v;
+    unw_get_reg(cursor, reg, &v);
+    return v;
+}
+static inline unw_word_t get_cursor_ip(unw_cursor_t* cursor) {
+    return get_cursor_reg(cursor, UNW_REG_IP) - 1;
+}
+static inline unw_word_t get_cursor_bp(unw_cursor_t* cursor) {
+    return get_cursor_reg(cursor, UNW_TDEP_BP);
+}
+
+template <typename FrameFunc>
+bool unwindProcessFrame(unw_word_t ip, unw_word_t bp, unw_cursor_t* cursor, FrameFunc func) {
+    CompiledFunction* cf = getCFForAddress(ip);
+    bool jitted = cf != NULL;
+    if (!cf) {
+        if (inASTInterpreterExecuteInner(ip)) {
+            cf = getCFForInterpretedFrame((void*)bp);
+            assert(cf);
+        }
+    }
+
+    if (!cf)
+        return false;
+
+    PythonFrameIteratorImpl info(jitted ? PythonFrameId::COMPILED : PythonFrameId::INTERPRETED, ip, bp, cf);
+    if (jitted) {
+        // Try getting all the callee-save registers, and save the ones we were able to get.
+        // Some of them may be inaccessible, I think because they weren't defined by that
+        // stack frame, which can show up as a -UNW_EBADREG return code.
+        for (int i = 0; i < 16; i++) {
+            if (!assembler::Register::fromDwarf(i).isCalleeSave())
+                continue;
+            unw_word_t r;
+            int code = unw_get_reg(cursor, i, &r);
+            ASSERT(code == 0 || code == -UNW_EBADREG, "%d %d", code, i);
+            if (code == 0) {
+                info.regs[i] = r;
+                info.regs_valid |= (1 << i);
+            }
+        }
+    }
+
+    if (threading::ThreadStateInternal::getUnwindState() == UNWIND_STATE_NORMAL) {
```

Comment:
I guess my point with both of the above comments is that `sys._getframe` has to deal with osr frames, and thus has to deal with the code to skip over a frame.  The only thing `sys._getframe` shouldn't have to deal with is the re-raise case.

By combining OSR and RERAISE into one state (since they both have the same meaning), we can remove the distasteful aspect of having `sys._getframe` care about re-raising.  It only cares about the possibility that the unwinding has to skip a frame (something it already had to deal with.)

This lets us reframe things:  The requirement is that `sys._getframe` shouldn't ever be used in a state where the *initial* frame is skipped.  That is already guaranteed by `unwindPythonStack` setting the state to `NORMAL` before walking the stack.  All of the entry points to unwinding that deal with introspection from python go through `unwindPythonStack`.

there are 4 places where we can begin unwinding the python stack, 3 of which originate in c++ unwinding.  we initialize the state in each of those places:

1. `raise0` - always sets the state to RERAISE/SKIPNEXT
2. `raise3` - sets the state to NORMAL or RERAISE/SKIPNEXT depending on arg3
3. `raiseExc` - always set the state to NORMAL (I actually need to add this)
3. `unwindPythonStack` - always sets the state to NORMAL


---

we can move the code to deal with maintaining the unwind state into the callers.  That simplifies unwindProcessFrame (now it will just create a PythonFrameIteratorImpl if it can, and call the function passed), and I think it's a good idea.  It just ends up duplicating the same state handling code in every use of `unwindPythonStack` except one: `getTopPythonFrame`.   The gist was just a way to move that code out that also won't c&p the same code everywhere.